### PR TITLE
Add a CEOD prober request handler to GAE and a cron job triggering the request regularly

### PR DESCRIPTION
**Pull request by jiayliu**

### CODE REVIEW

File: samples/web/content/apprtc/constants.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+"""AppRTC Constants
+
+This module contains the constants used in AppRTC Python modules.
+"""
+
+LOOPBACK_CLIENT_ID = 'LOOPBACK_CLIENT_ID'
+
+TURN_BASE_URL = 'https://computeengineondemand.appspot.com'
+TURN_URL_TEMPLATE = '%s/turn?username=%s&key=%s'
```
Comment (juberti) [inline samples/web/content/apprtc/constants.py]: call this TURN_PATH_TEMPLATE

File: samples/web/content/apprtc/constants.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+CEOD_KEY = '4080218913'
+
+WSS_HOST = 'apprtc-ws.webrtc.org'
```
Comment (juberti) [inline samples/web/content/apprtc/constants.py]: remind me, why can't we just store this as a single constant WSS_URL (and have a single param to overrride it)?

File: samples/web/content/apprtc/probers.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+         dict[key] != ''
+
+def has_non_empty_array_value(dict, key):
+  return key in dict and \
+         isinstance(dict[key], list) and \
+         len(dict[key]) > 0
+
+class ProbeCEODPage(webapp2.RequestHandler):
+  def get(self):
+    alert_email = self.request.get('email')
```
Comment (juberti) [inline samples/web/content/apprtc/probers.py]: allowing an external user to (indirectly) send email seems dangerous.

File: samples/web/content/apprtc/probers.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    ceod_url = constants.TURN_URL_TEMPLATE \
+        % (constants.TURN_BASE_URL, 'prober', constants.CEOD_KEY)
+    insensitive_url = constants.TURN_URL_TEMPLATE % \
```
Comment (juberti) [inline samples/web/content/apprtc/probers.py]: redacted_url or sanitized_url is the typical way to name this

File: samples/web/content/apprtc/probers.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    if status_code != 200:
+      error_message = 'Unexpected CEOD response: %d, requested URL: %s' \
+          % (result.status_code, insensitive_url)
+    else:
+      try:
+        turn_server = json.loads(result.content)
+        if not has_non_empty_string_value(turn_server, 'username') or \
+           not has_non_empty_string_value(turn_server, 'password') or \
+           not has_non_empty_array_value(turn_server, 'uris'):
+          error_message = 'CEOD response does not contain valid ' + \
```
Comment (juberti) [inline samples/web/content/apprtc/probers.py]: include URL?

File: samples/web/content/apprtc/probers.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+              'username/password/uris: ' + result.content
+          status_code = 500
+      except Exception as e:
+        error_message = """
```
Comment (juberti) [inline samples/web/content/apprtc/probers.py]: include URL?

File: samples/web/content/apprtc/probers.py (hunk 3 above)
Comment (juberti) [inline samples/web/content/apprtc/probers.py]: that is, to an arbitrary user. I think we probably want to have the target email address stored in the code or in a DB somewhere, not passed in from the URL.

File: samples/web/content/apprtc/probers.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+"""
+
+import constants
+import logging
+import json
+import webapp2
+from google.appengine.api import mail
+from google.appengine.api import urlfetch
+
+def sendAlertEmail(receiver, tag, message):
```
Comment (juberti) [inline samples/web/content/apprtc/probers.py]: send_alert_email is python style, IIRC

File: samples/web/content/apprtc/constants.py (hunk 1 above)
Comment (jiayliu) [inline samples/web/content/apprtc/constants.py]: It's the template for the full URL, not just the path.

**Comment by jiayliu** [comment]
New change uploaded. PTAL.
The probe request is now limited to 'admin', which cron jobs can run.

File: samples/web/content/apprtc/constants.py (hunk 1 above)
Comment (juberti) [inline samples/web/content/apprtc/constants.py]: OK

File: samples/web/content/apprtc/constants.py (hunk 2 above)
Comment (juberti) [inline samples/web/content/apprtc/constants.py]: ping

**Comment by juberti** [comment]
lgtm

re: WSS_HOST: remind me, why can't we just store this as a single constant WSS_URL (and have a single param to overrride it)?

**Comment by jiayliu** [comment]
We needs both the wss:// url and the https:// url composed from
wss_host_port_pair.

On Fri, Jan 9, 2015 at 1:20 PM, Justin Uberti <notifications@github.com>
wrote:

> lgtm
>
> re: WSS_HOST: remind me, why can't we just store this as a single constant
> WSS_URL (and have a single param to overrride it)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleChrome/webrtc/pull/333#issuecomment-69400705>.
>



-- 
Jiayang Liu

**Pull request merged**